compiler/rustc_mir_build/src/check_tail_calls.rs

```diff
@@ -280,6 +287,31 @@ impl<'tcx> TailCallCkVisitor<'_, 'tcx> {
         self.found_errors = Err(err);
     }
 
+    fn report_nonfn_callee(&mut self, call_sp: Span, fun_sp: Span, ty: Ty<'_>) {
+        let mut err = self
+            .tcx
+            .dcx()
+            .struct_span_err(
+                call_sp,
+                "tail calls can only be performed with function definitions or pointers",
+            )
+            .with_note(format!("callee has type `{ty}`"));
+
+        if ty.is_ref() {
```

Comment:
Given that the suggestion is MachineApplicable, we should either actually check that the dereference will get a function definition or not mark it as MachineApplicable.


```suggestion
        if ty::Ref(_, inner_ty, _) = ty.kind() && inner_ty.is_fn() {
```

Reply:
I don't think it's explicit but I've always followed that a machine-applicable suggestion should always result in working code. I'm pretty sure `cargo fix`/`rustfix` expects working code for machine-applicable suggestions.

When we are not 100% certain it's going to fix the issue, even if we know a 100% it will lead to working code, we should IMO use `MaybeIncorrect`, and that's what I have been following.